openfisca_core/entities.py

```diff
@@ -276,24 +199,72 @@ def get_rank(self, entity, criteria, condition = True):
         return np.where(condition, result, -1)
 
 
-class GroupEntity(Entity):
+class Entity(object):
     """
-        Represents an entity composed of several persons with different roles, on which calculations are run.
+        Represents an entity (e.g. a person, a household, etc.) on which calculations can be run.
     """
+    def __init__(self, key, plural, label, doc):
+        self.key = key
+        self.label = label
+        self.plural = plural
+        self.doc = textwrap.dedent(doc)
+        self.is_person = True
+        self.tax_benefit_system = None
+
+    def set_tax_benefit_system(self, tax_benefit_system):
```

Comment:
`check_variable_defined_for_entity` is also used in `__call__`. This error typically happens if I try to get `person('rent', period)` while `rent` is defined for a `household`.

Reply:
Ok, I think I understand the problem better, and indeed having `Entity` keep a reference to the TaxBenefitSystem doesn't feel great at all.
I'm however always a little worried about "temporarily" making changes to the codebase that we don't find satisfying. It's hard to predict when we'll actually come back to it... 

Could we try to find a way to move `check_variable_defined_for_entity` and `get_variable` (which seems to be the core of the problem) out of `Entity` ? `get_variable` just looks up a variable in the TBS by its name, which is not really related to `Entity`.

Reply:
> check_variable_defined_for_entity is also used in __call__. This error typically happens if I try to get person('rent', period) while rent is defined for a household.

Yes, and that seems a reasonable precaution. It's less clear why `get_holder` needs it.

> I'm however always a little worried about "temporarily" making changes to the codebase that we don't find satisfying.

You and me both. :) However this PR seems to me a strict improvement on what came before, where the corresponding navigation was in class Entity as follows:
```
def get_variable(self, variable_name, check_existence = False):
  return self.simulation.tax_benefit_system.get_variable(variable_name, check_existence)
```

In calls like these, and whether `person` is a population or an entity:
```
person('rent', period)
```
the only way to know about the variable-entity link is to ask the TaxBenefitSystem. (A possible evolution would be to make each Entity responsible for keeping track of its own variables, in the same way that each Population is keeping track of its own values; but that's another major restructuring, and I'd argue well out of scope for this PR.)

To get a TaxBenefitSystem instance, either we explicitly hold a reference to it - which is what this PR does - or we use the one in the Simulation instance we have for almost entirely different reasons (i.e. so we can call `calculate`) - this is what was happening previously; it's a Demeter violation and has a "sneaky" feel to it.

This is related to the issue we had with the period check; passing in a Simulation to the formulas, so that they can `calculate` with it, makes sense because Simulation has a "bird's eye view" of the system; passing in an Entity (or as is now the case a Population) makes less sense, because these objects have a narrower view and are not able to do the job by themselves.

I understand that we might not want to go back to the previous formula signature where Simulation was passed in. (I'd still be curious to know more about the reasons for the change, if you remember the PRs in which it was done and where discussions might have taken place.)

And to me there is a different way forward, which would eliminate a lot of boilerplate calls from formulas, unlock a bunch of optimizations, and facilitate static dependency computation; in a nutshell, it consists of reintroducing the variable dependencies as parameters of the formulas. But that's a bigger task and needs a separate RFC, discussion and plan.

Reply:
> I'd still be curious to know more about the reasons

Nevermind, found #408.

Reply:
>  consists of reintroducing the variable dependencies as parameters of the formulas. 

I think that how OpenFisca worked in the very beginning, but had to be dropped to handle  periods and entities.

Reply:
Anyway, I don't have a better solution to suggest here, so 👌  for me.

Reply:
> how OpenFisca worked in the very beginning, but had to be dropped to handle periods and entities

Yes, I found the commit - 1b17255034 - which suggests that the reason was to handle varying periods.